File: chapter 09/recipe 9.1/DniScMonitor/DniScMonitorObj.cs
using System;
using System.Runtime.InteropServices;
using System.EnterpriseServices;

using TransactionLogging;

namespace DniScMonitor
{
	public interface ITranMonitor
	{
		void LogTransactionDetails();
	}

	[Transaction(TransactionOption.Required)]
	[ClassInterface(ClassInterfaceType.None)]
	public class DniScMonitorObj
		: ServicedComponent, ITranMonitor
	{
		public void LogTransactionDetails()
		{
			//create logging object that will record
			//the transaction state
			TransactionLogger logger
				= new TransactionLogger(this);
		}
	}
}

File: chapter 09/recipe 9.1/DniScMonitorVB/DniScMonitorObj.vb
Imports System.EnterpriseServices
Imports System.Transactions
Imports System.Runtime.InteropServices

Public Interface ITranMonitor
    Sub LogTransactionDetails()
End Interface

<Transaction(TransactionOption.Required)> _
<ClassInterface(ClassInterfaceType.None)> _
Public Class DniScMonitorObj
    Inherits ServicedComponent
    Implements ITranMonitor

    Public Sub LogTransactionDetails() _
        Implements ITranMonitor.LogTransactionDetails

        'create the logging class
        Dim logger As TransactionLogging.TransactionLogger _
           = New TransactionLogging.TransactionLogger(Me)

    End Sub
End Class
